### PR TITLE
Perspective correction

### PR DESCRIPTION
**Pull request by Mati-AAU**

### CODE REVIEW

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/GLOutputStage.kt.orig [hunk 1]
```diff
@@ -0,0 +1,175 @@
+package dk.scuffed.whiteboardapp.pipeline.stages
```
Comment (mads256h) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/GLOutputStage.kt.orig:1]: Delete this file

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/PerspectiveCorrectionStage.kt [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        return if (determinant == 0.0f) {
+            // The lines are parallel. This is simplified
+            // by returning a pair of FLT_MAX
+            Vec2Float(Float.MAX_VALUE, Float.MAX_VALUE)
+        } else {
+            val x = (b2 * c1 - b1 * c2) / determinant
+            val y = (a1 * c2 - a2 * c1) / determinant
+            Vec2Float(x, y)
+        }
+    }
```
Comment (mads256h) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/PerspectiveCorrectionStage.kt:128]: Use the function in LineFloat

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        )
+
+        var dst: MatOfPoint2f? = MatOfPoint2f(
+            Point(0, 0),
+            Point(450 - 1, 0),
+            Point(0, 450 - 1),
+            Point(450 - 1, 450 - 1)
+        )        Imgproc.getPerspectiveTransform()
+        val transformMatrix = Core.p
+        */
```
Comment (mads256h) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt:73]: Remove these comments

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+
+
+        val newpoints = result.toArray()
+
+
+        for (i in points.indices) {
+            points[i] = Vec2Int(newpoints[i].x.toInt(), newpoints[i].y.toInt())
+        }
+
```
Comment (mads256h) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt:57]: Free all your mats here

**Review by mads256h — CHANGES_REQUESTED** [review]

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/PerspectiveCorrectionStage.kt (hunk 2 above)
Comment (Mati-AAU) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/PerspectiveCorrectionStage.kt]: done

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt (hunk 4 above)
Comment (Mati-AAU) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt:58]: done

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/GLOutputStage.kt.orig (hunk 1 above)
Comment (Mati-AAU) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/GLOutputStage.kt.orig]: done

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            )
+
+        val dst = MatOfPoint2f(
+            Point(pointsTo.points[0].x.toDouble(), pointsTo.points[0].y.toDouble()),
+            Point(pointsTo.points[1].x.toDouble(), pointsTo.points[1].y.toDouble()),
+            Point(pointsTo.points[2].x.toDouble(), pointsTo.points[2].y.toDouble()),
+            Point(pointsTo.points[3].x.toDouble(), pointsTo.points[3].y.toDouble()),
+        )
+
+        val warpmat = Imgproc.getPerspectiveTransform(src, dst)
```
Comment (S3JER) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/points_stages/PerspectiveTransformPointsStage.kt:47]: Maybe a new name to warpmat

**Review by S3JER — CHANGES_REQUESTED** [review]
Only some places need docs, else LGTM

**Review by S3JER — APPROVED** [review]
LGTM

**Review by S3JER — APPROVED** [review]

**Pull request merged**